File: app/src/main/java/com/pr0gramm/app/ui/fragments/InboxFragment.java
package com.pr0gramm.app.ui.fragments;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import com.pr0gramm.app.R;
import com.pr0gramm.app.api.pr0gramm.response.Message;
import com.pr0gramm.app.feed.FeedType;
import com.pr0gramm.app.services.InboxService;
import com.pr0gramm.app.services.UriHelper;
import com.pr0gramm.app.ui.InboxType;
import com.pr0gramm.app.ui.MainActivity;
import com.pr0gramm.app.ui.MessageActionListener;
import com.pr0gramm.app.ui.WriteMessageActivity;
import com.pr0gramm.app.ui.base.BaseFragment;
import com.squareup.picasso.Picasso;

import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;

import static com.pr0gramm.app.ui.dialogs.ErrorDialogFragment.defaultOnError;
import static org.joda.time.Duration.standardMinutes;

/**
 */
public abstract class InboxFragment<T> extends BaseFragment {
    protected final Logger logger = LoggerFactory.getLogger(getClass());

    protected static final String ARG_INBOX_TYPE = "InboxFragment.inboxType";

    @Inject
    InboxService inboxService;

    @Inject
    Picasso picasso;

    @Bind(R.id.messages)
    RecyclerView messagesView;

    @Bind(android.R.id.empty)
    View viewNothingHere;

    @Bind(R.id.busy_indicator)
    View viewBusyIndicator;

    @Bind(R.id.refresh)
    SwipeRefreshLayout swipeRefreshLayout;

    private LoaderHelper<List<T>> loader;
    private Instant loadStartedTimestamp;

    public InboxFragment() {
        setRetainInstance(true);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.loader = newLoaderHelper();
        this.loader.reload();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_inbox, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        messagesView.setItemAnimator(null);
        messagesView.setLayoutManager(new LinearLayoutManager(getActivity()));

        swipeRefreshLayout.setOnRefreshListener(this::reloadInboxContent);
        swipeRefreshLayout.setColorSchemeResources(R.color.orange_primary);

        showBusyIndicator();

        // load the messages
        loadStartedTimestamp = Instant.now();
        loader.load(this::onMessagesLoaded, defaultOnError(), this::hideBusyIndicator);
    }

    @Override
    public void onResume() {
        super.onResume();

        // reload if re-started after one minute
        if (loadStartedTimestamp.plus(standardMinutes(1)).isBeforeNow()) {
            loadStartedTimestamp = Instant.now();
            reloadInboxContent();
        }
    }

    @Override
    public void onDestroyView() {
        loader.detach();
        super.onDestroyView();
    }

    private void reloadInboxContent() {
        loader.reload();
    }

    private void hideNothingHereIndicator() {
        if (hasView()) {
            viewNothingHere.setVisibility(View.GONE);
        }
    }

    private void showNothingHereIndicator() {
        if (hasView()) {
            viewNothingHere.setVisibility(View.VISIBLE);
            viewNothingHere.setAlpha(0);
            viewNothingHere.animate().alpha(1).start();
        }
    }

    private void showBusyIndicator() {
        if (hasView() && viewBusyIndicator != null) {
            viewBusyIndicator.setVisibility(View.VISIBLE);
        }
    }

    private void hideBusyIndicator() {
        if (hasView()) {
            if (viewBusyIndicator != null) {
                viewBusyIndicator.setVisibility(View.GONE);
                ViewParent parent = viewBusyIndicator.getParent();
                ((ViewGroup) parent).removeView(viewBusyIndicator);

                viewBusyIndicator = null;
            }

            swipeRefreshLayout.setRefreshing(false);
        }
    }

    protected abstract LoaderHelper<List<T>> newLoaderHelper();

    private boolean hasView() {
        return messagesView != null;
    }

    private void onMessagesLoaded(List<T> messages) {
        hideBusyIndicator();
        hideNothingHereIndicator();

        // replace previous adapter with new values
        displayMessages(messagesView, messages);

        if (messages.isEmpty())
            showNothingHereIndicator();
    }

    protected abstract void displayMessages(RecyclerView recyclerView, List<T> messages);

    protected InboxType getInboxType() {
        InboxType type = InboxType.ALL;
        Bundle args = getArguments();
        if (args != null) {
            type = InboxType.values()[args.getInt(ARG_INBOX_TYPE, InboxType.ALL.ordinal())];
        }

        return type;
    }

    protected InboxService getInboxService() {
        return inboxService;
    }

    protected final MessageActionListener actionListener = new MessageActionListener() {
        @Override
        public void onAnswerToPrivateMessage(Message message) {
            startActivity(WriteMessageActivity.intent(getActivity(), message));
        }

        @Override
        public void onAnswerToCommentClicked(Message message) {
            startActivity(WriteMessageActivity.answerToComment(getActivity(), message));
        }

        @Override
        public void onNewPrivateMessage(long userId, String name) {
            startActivity(WriteMessageActivity.intent(getActivity(), userId, name));
        }

        @Override
        public void onCommentClicked(long itemId, long commentId) {
            open(UriHelper.of(getActivity()).post(FeedType.NEW, itemId, commentId));
        }

        private void open(Uri uri) {
            Intent intent = new Intent(Intent.ACTION_VIEW, uri, getActivity(), MainActivity.class);
            startActivity(intent);
        }

        @Override
        public void onUserClicked(int userId, String username) {
            open(UriHelper.of(getActivity()).uploads(username));
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/ui/fragments/InboxFragment.kt
package com.pr0gramm.app.ui.fragments

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.pr0gramm.app.Instant
import com.pr0gramm.app.R
import com.pr0gramm.app.api.pr0gramm.Message
import com.pr0gramm.app.databinding.FragmentInboxBinding
import com.pr0gramm.app.feed.FeedType
import com.pr0gramm.app.services.InboxService
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.UriHelper
import com.pr0gramm.app.ui.ConversationActivity
import com.pr0gramm.app.ui.MainActivity
import com.pr0gramm.app.ui.MarginDividerItemDecoration
import com.pr0gramm.app.ui.MessageActionListener
import com.pr0gramm.app.ui.MessageAdapter
import com.pr0gramm.app.ui.Pagination
import com.pr0gramm.app.ui.WriteMessageActivity
import com.pr0gramm.app.ui.addEndStateToValues
import com.pr0gramm.app.ui.base.BaseFragment
import com.pr0gramm.app.ui.base.bindViews
import com.pr0gramm.app.ui.fragments.conversation.StringValue
import com.pr0gramm.app.util.di.instance
import com.pr0gramm.app.util.observeChange
import java.util.concurrent.TimeUnit

/**
 */
abstract class InboxFragment(name: String) : BaseFragment(name, R.layout.fragment_inbox) {
    protected val inboxService: InboxService by instance()

    private val views by bindViews(FragmentInboxBinding::bind)

    private var loadStartedTimestamp = Instant(0)

    private lateinit var adapter: MessageAdapter
    private lateinit var pagination: Pagination<Message>

    private var state by observeChange(State()) { updateAdapterValues() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val (adapter, pagination) = getContentAdapter()
        this.adapter = adapter
        this.pagination = pagination
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        views.messages.itemAnimator = null
        views.messages.layoutManager = LinearLayoutManager(activity)
        views.messages.adapter = adapter
        views.messages.addItemDecoration(MarginDividerItemDecoration(requireContext(), marginLeftDp = 72))

        views.refresh.setOnRefreshListener { reloadInboxContent() }
        views.refresh.setColorSchemeResources(ThemeHelper.accentColor)

        pagination.updates.observe(viewLifecycleOwner) { (state, newValues) ->
            handleStateUpdate(state, newValues)
        }
    }

    override fun onResume() {
        super.onResume()

        // reload if re-started after one minute
        if (loadStartedTimestamp.plus(1, TimeUnit.MINUTES).isBefore(Instant.now())) {
            loadStartedTimestamp = Instant.now()
            reloadInboxContent()
        }
    }

    private fun reloadInboxContent() {
        views.refresh.isRefreshing = false

        // re-set state and re-start pagination
        state = State()
        pagination.initialize()
    }

    abstract fun getContentAdapter(): Pair<MessageAdapter, Pagination<Message>>

    private fun handleStateUpdate(state: Pagination.State<Message>, newValues: List<Message>) {
        this.state = this.state.copy(
                messages = (this.state.messages + newValues).distinctBy { msg -> msg.id },
                tailState = state.tailState)
    }

    private fun updateAdapterValues() {
        val context = context ?: return

        // build values for the recycler view
        val values = state.messages.toMutableList<Any>()

        // add the divider above the first read message, but only
        // if we have at least one unread message
        val dividerIndex = values.indexOfFirst { it is Message && it.read }
        if (dividerIndex > 0) {
            val divider = StringValue(context.getString(R.string.inbox_type_unread))
            values.add(dividerIndex, divider)
        }

        // add common state values
        addEndStateToValues(context, values, state.tailState, ifEmptyValue = MessageAdapter.EmptyValue)

        adapter.submitList(values)
    }

    protected val actionListener: MessageActionListener = object : MessageActionListener {
        override fun onAnswerToPrivateMessage(message: Message) {
            ConversationActivity.start(context ?: return, message.name, skipInbox = true)
        }

        override fun onAnswerToCommentClicked(comment: Message) {
            startActivity(WriteMessageActivity.answerToComment(context ?: return, comment))
        }

        override fun onNewPrivateMessage(userId: Long, name: String) {
            startActivity(WriteMessageActivity.intent(context ?: return, userId, name))
        }

        override fun onCommentClicked(comment: Message) {
            val uri = UriHelper.of(context ?: return).post(FeedType.NEW, comment.itemId, comment.id)
            open(uri, comment.creationTime)
        }

        private fun open(uri: Uri, notificationTime: Instant? = null) {
            val intent = Intent(Intent.ACTION_VIEW, uri, context, MainActivity::class.java)
            intent.putExtra("MainActivity.NOTIFICATION_TIME", notificationTime)
            startActivity(intent)
        }

        override fun onUserClicked(userId: Int, username: String) {
            open(UriHelper.of(context ?: return).uploads(username))
        }
    }

    data class State(
            val messages: List<Message> = listOf(),
            val tailState: Pagination.EndState<Message> = Pagination.EndState(hasMore = true))
}
